File: Qytetet_el_juego/p4/QytetetJava/src/modeloqytetet/Calle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author mati
 */
public class Calle extends Casilla {
    
    private int numHoteles;
    private int numCasas;
    private TituloPropiedad titulo;
    
    public Calle(int numeroCasilla, int coste, int numHoteles, int numCasas, TipoCasilla tipo, TituloPropiedad titulo) {
        super(numeroCasilla,coste,tipo);
        this.numHoteles = numHoteles;
        this.numCasas = numCasas;
        this.titulo = titulo;
        this.titulo.setCasilla(this);       
    }
    
    public TituloPropiedad getTituloPropiedad() {
        return titulo;
    }
    
    TituloPropiedad asignarPropietario( Jugador jugador){
        this.titulo.setPropietario(jugador);
        return titulo ;
    }
    
    int calcularValorHipoteca(){
        int cantidad = (this.titulo.getHipotecaBase()+ (this.numCasas * 5/10 * this.titulo.getHipotecaBase()+ (this.numHoteles * this.titulo.getAlquilerBase())));
        return cantidad ;
    }
    
    int cancelarHipoteca(){ 
        return (this.calcularValorHipoteca()*10/100) + this.calcularValorHipoteca();
    }
    
    int cobrarAlquiler(){ 
        int costeAlquilerBase = this.titulo.getAlquilerBase();
        int coste = (costeAlquilerBase + (2 *((this.numCasas * 10/5) + (this.numHoteles * 2)))) ;
        this.titulo.cobrarAlquiler(coste);
        return coste;
    }
    
    int edificarCasa(){
        this.setNumCasas(this.numCasas + 1);
        int costeEdificarCasa = this.getPrecioEdificar();
        return costeEdificarCasa ;
    }
    
    int edificarHotel(){
        this.setNumHoteles(numHoteles + 1 );
        int costeEdificarHotel = this.getPrecioEdificar();
        return costeEdificarHotel ;
    }
    
    boolean estaHipotecada(){
        return this.titulo.getHipotecada();
    } 
    
    int getCosteHipoteca(){
        return this.calcularValorHipoteca();
    }
    
    int getNumHoteles() {
        return numHoteles;
    }
    
    int getNumCasas() {
        return numCasas;
    }
    
    int getPrecioEdificar(){
        return this.titulo.getPrecioEdificar() ;
    }
   
    int hipotecar(){
        this.titulo.setHipotecada(true);
        return this.getCosteHipoteca();
    }
    
    int precioTotalComprar(){
        return this.titulo.getAlquilerBase();
    }
    
    boolean propietarioEncarcelado(){
        return this.titulo.propietaroEncarcelado();
    }
    
    public boolean sePuedeEdificarCasa(int factorEspeculador){
        boolean sepuede = false ;
        if(this.numCasas < (4 * factorEspeculador) && this.soyEdificable()){
            sepuede = true ;
        }
        return sepuede ;
    }
    
    public boolean sePuedeEdificarHotel(int factorEspeculador){
        boolean sepuede = false ;
        if(this.numHoteles < (4 * factorEspeculador) && soyEdificable() ){
            sepuede = true ;
        }
        return sepuede ;
    }
    
    void setNumHoteles(int numHoteles) {
        this.numHoteles = numHoteles;
    }
    
    void setNumCasas(int numCasas) {
        this.numCasas = numCasas;
    }
    
    public boolean tengoPropietario(){ // cambio de package a publico
        return this.titulo.tengoPropietario();
    }
    
    int venderTitulo(){
        int precio = this.precioTotalComprar();
        precio += this.numHoteles * this.getPrecioEdificar();
        precio += this.numCasas * this.getPrecioEdificar();
        precio *= this.titulo.getFactorDevolucion();
        this.titulo.setPropietario(null);
        this.setNumCasas(0);
        this.setNumHoteles(0);
        return precio ;
    }
    
    
   // @Override
    void setTituloPropiedad(TituloPropiedad titulo) {
        this.titulo = titulo;
    }
    
   // @Override
    void asignarTituloPropiedad(){
        this.titulo = this.getTituloPropiedad() ;
    }
    
}

File: Qytetet_el_juego/p4/QytetetRuby/lib/calle.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class Calle < Casilla
        attr_accessor :numHoteles, :numCasas, :titulo, :coste
       
    def initialize(numCasilla,cos, titul)  
      super(numCasilla, TipoCasilla::CALLE)
      @coste = cos
      @numHoteles = 0 # int
      @numCasas = 0  # int
      @titulo = titul  # TituloPropiedad
      if @titulo != nil 
        @titulo.setCasilla(self) 
      end
    end
    
    def asignarPropietario(jugador) #TituloPropiedad asignarPropietario( Jugador jugador)
      @titulo.setPropietario(jugador)
    end
    
    def calcularValorHipoteca #int calcularValorHipoteca()
      return (@titulo.hipotecaBase + 2 * (@numCasas * 5/10 * @titulo.hipotecaBase + @numHoteles + @titulo.hipotecaBase ))
    end
    
    def cancelarHipoteca #int cancelarHipoteca() #***************************************************************  @titulo.hipotecada = false
      return (calcularValorHipoteca()*10/100) + calcularValorHipoteca();
    end
    
    def cobrarAlquiler #int cobrarAlquiler()
      costeAlquilerBase = @titulo.alquilerBase
      coste = (costeAlquilerBase + (2 *((@numCasas * 5/10) + (@numHoteles * 2))))
      @titulo.cobrarAlquiler(coste)
      return coste
    end
    
    def edificarCasa #int edificarCasa()
      @numCasas += 1
      return @titulo.precioEdificar
    end
    
    def edificarHotel #int edificarHotel()
      @numHoteles += 1
      return @titulo.precioEdificar
    end
    
    def estaHipotecada #boolean estaHipotecada()
      return @titulo.hipotecada
    end
    
    def getPrecioEdificar #int getPrecioEdificar()
      @titulo.precioEdificar
    end
    
    def hipotecar #int hipotecar()
      @titulo.hipotecada=true ##############################
      cantidadRecibida = calcularValorHipoteca
      return cantidadRecibida
    end
    
    def precioTotalComprar #int precioTotalComprar()
      @titulo.alquilerBase
    end
    
    def propietarioEncarcelado #boolean propietarioEncarcelado()
      @titulo.propietaroEncarcelado
    end
    
    def sePuedeEdificarCasa(factor) #boolean sePuedeEdificarCasa()
      return (@numCasas < 4 * factor)
    end
    
    def sePuedeEdificarHotel(factor) #boolean sePuedeEdificarHotel()
      return (@numHoteles < 4 * factor)
    end
    
    def tengoPropietario #boolean tengoPropietario() 
      @titulo.tengoPropietario
    end
    
    def venderTitulo #int venderTitulo()
      precio = precioTotalComprar
      precio += @numHoteles * getPrecioEdificar
      precio += @numCasas * getPrecioEdificar
      precio *= @titulo.factorRevalorizacion
      asignarPropietario(nil)
      @numHoteles = 0
      @numCasas = 0
      return precio
    end
    
    private
    
    def asignarTituloPropiedad #private void asignarTituloPropiedad()
      raise ArgumentError,'metodo sin implementar'
    end
    
    public
    
    def to_s
      cadena = super.to_s
      cadena +=" CALLE:  
      numHoteles: #{@numHoteles} 
      numCasas: #{@numCasas}  \n "
      if(@titulo != nil)
        cadena +=" tituloDeCasilla: #{@titulo.to_s} "
      else
        cadena +=" tituloDeCasilla: no tiene titulo"
      end
      cadena += " \nFin de casilla\n"
      return cadena
    end  
  end
end
